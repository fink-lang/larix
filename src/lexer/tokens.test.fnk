{describe, it, expect, to_match_snapshot} = import '@fink/jest/test.fnk'
{obj_to_json} = import '@fink/std-lib/json.fnk'
{join} = import '@fink/std-lib/iter.fnk'

{tokenize} = import './tokens.fnk'



foo = fn source:
  pipe source:
    tokenize
    until token: token.type == 'end'
    map {type, value, loc: {start, end}}:
      pos = '(${start.pos}-${end.pos})'
      loc = '(${start.line}:${start.column}-${end.line}:${end.column})'
      match value:
        type:
          '${type} ${pos} ${loc}'
        else:
          '${type} ${pos} ${loc}\n  ${obj_to_json value}'

    join '\n'



describe 'tokenizer', fn:

  it 'tokenizes', fn:
    expect
      foo '
        01234567
        foo bar
        (shrub)
        (spam ni, hi)
        a + b - c * d / f ^ g
        "shrub"
        "ni \${ham} spam"
        foo.bar ...ni
      '
      to_match_snapshot


  it 'tokenizes str exprs', fn:
    expect
      foo '
        {foo: "\${ham}"}
      '
      to_match_snapshot



describe 'JSX tokenizer', fn:
  it 'tokenizes fragment', fn:
    expect
      foo '
        <>
          foo bar
          <foo>
            spam {1 + 2}
          </foo>
        </>
      '
      to_match_snapshot


  it 'tokenizes jsx', fn:
    expect
      foo '
        <foo>{a > b, {a: 12345}, "foo \${ni} bar"}</foo>
      '
      to_match_snapshot

    expect
      foo '
        <>
          <foo />
        </>
      '
      to_match_snapshot


  it 'tokenizes expr-attr', fn:
    expect
      foo '
        <Foobar spam={ni} />
      '
      to_match_snapshot
