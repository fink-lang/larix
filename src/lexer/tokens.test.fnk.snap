// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSX tokenizer tokenizes expr-attr 1`] = `
"jsx-elem-start (0-1) (1:0-1:1)
  \\"<\\"
ident (1-7) (1:1-1:7)
  \\"Foobar\\"
ignorable (7-8) (1:7-1:8)
  \\" \\"
ident (8-12) (1:8-1:12)
  \\"spam\\"
= (12-13) (1:12-1:13)
jsx-expr-start (13-14) (1:13-1:14)
  \\"{\\"
ident (14-16) (1:14-1:16)
  \\"ni\\"
jsx-expr-end (16-17) (1:16-1:17)
  \\"}\\"
ignorable (17-18) (1:17-1:18)
  \\" \\"
jsx-elem-close (18-20) (1:18-1:20)
  \\"/>\\"
ignorable (20-21) (1:20-2:0)
  \\"\\\\n\\"
end (21-21) (2:0-2:0)
  \\"\\""
`;

exports[`JSX tokenizer tokenizes fragment 1`] = `
"jsx-frag-open (0-2) (1:0-1:2)
  \\"<>\\"
jsx-text (2-15) (1:2-3:2)
  \\"\\\\n  foo bar\\\\n  \\"
jsx-elem-start (15-16) (3:2-3:3)
  \\"<\\"
ident (16-19) (3:3-3:6)
  \\"foo\\"
jsx-elem-end (19-20) (3:6-3:7)
  \\">\\"
jsx-text (20-30) (3:7-4:9)
  \\"\\\\n    spam \\"
jsx-expr-start (30-31) (4:9-4:10)
  \\"{\\"
number (31-32) (4:10-4:11)
  \\"1\\"
ignorable (32-33) (4:11-4:12)
  \\" \\"
+ (33-34) (4:12-4:13)
ignorable (34-35) (4:13-4:14)
  \\" \\"
number (35-36) (4:14-4:15)
  \\"2\\"
jsx-expr-end (36-37) (4:15-4:16)
  \\"}\\"
jsx-text (37-40) (4:16-5:2)
  \\"\\\\n  \\"
jsx-elem-close (40-46) (5:2-5:8)
  \\"</foo>\\"
jsx-text (46-47) (5:8-6:0)
  \\"\\\\n\\"
jsx-elem-close (47-50) (6:0-6:3)
  \\"</>\\"
ignorable (50-51) (6:3-7:0)
  \\"\\\\n\\"
end (51-51) (7:0-7:0)
  \\"\\""
`;

exports[`JSX tokenizer tokenizes hyphenated attr 1`] = `
"jsx-elem-start (0-1) (1:0-1:1)
  \\"<\\"
ident (1-7) (1:1-1:7)
  \\"Foobar\\"
ignorable (7-8) (1:7-1:8)
  \\" \\"
ident (8-16) (1:8-1:16)
  \\"spam-ham\\"
ignorable (16-17) (1:16-1:17)
  \\" \\"
jsx-elem-close (17-19) (1:17-1:19)
  \\"/>\\"
ignorable (19-20) (1:19-2:0)
  \\"\\\\n\\"
end (20-20) (2:0-2:0)
  \\"\\""
`;

exports[`JSX tokenizer tokenizes jsx 1`] = `
"jsx-elem-start (0-1) (1:0-1:1)
  \\"<\\"
ident (1-4) (1:1-1:4)
  \\"foo\\"
jsx-elem-end (4-5) (1:4-1:5)
  \\">\\"
jsx-expr-start (5-6) (1:5-1:6)
  \\"{\\"
ident (6-7) (1:6-1:7)
  \\"a\\"
ignorable (7-8) (1:7-1:8)
  \\" \\"
> (8-9) (1:8-1:9)
ignorable (9-10) (1:9-1:10)
  \\" \\"
ident (10-11) (1:10-1:11)
  \\"b\\"
, (11-12) (1:11-1:12)
ignorable (12-13) (1:12-1:13)
  \\" \\"
{ (13-14) (1:13-1:14)
ident (14-15) (1:14-1:15)
  \\"a\\"
: (15-16) (1:15-1:16)
ignorable (16-17) (1:16-1:17)
  \\" \\"
number (17-22) (1:17-1:22)
  \\"12345\\"
} (22-23) (1:22-1:23)
, (23-24) (1:23-1:24)
ignorable (24-25) (1:24-1:25)
  \\" \\"
str-start (25-26) (1:25-1:26)
  \\"\\\\\\"\\"
str-text (26-30) (1:26-1:30)
  \\"foo \\"
str-expr-start (30-32) (1:30-1:32)
  \\"\${\\"
ident (32-34) (1:32-1:34)
  \\"ni\\"
str-expr-end (34-35) (1:34-1:35)
  \\"}\\"
str-text (35-39) (1:35-1:39)
  \\" bar\\"
str-end (39-40) (1:39-1:40)
  \\"\\\\\\"\\"
jsx-expr-end (40-41) (1:40-1:41)
  \\"}\\"
jsx-elem-close (41-47) (1:41-1:47)
  \\"</foo>\\"
ignorable (47-48) (1:47-2:0)
  \\"\\\\n\\"
end (48-48) (2:0-2:0)
  \\"\\""
`;

exports[`JSX tokenizer tokenizes jsx 2`] = `
"jsx-frag-open (0-2) (1:0-1:2)
  \\"<>\\"
jsx-text (2-5) (1:2-2:2)
  \\"\\\\n  \\"
jsx-elem-start (5-6) (2:2-2:3)
  \\"<\\"
ident (6-9) (2:3-2:6)
  \\"foo\\"
ignorable (9-10) (2:6-2:7)
  \\" \\"
jsx-elem-close (10-12) (2:7-2:9)
  \\"/>\\"
jsx-text (12-13) (2:9-3:0)
  \\"\\\\n\\"
jsx-elem-close (13-16) (3:0-3:3)
  \\"</>\\"
ignorable (16-17) (3:3-4:0)
  \\"\\\\n\\"
end (17-17) (4:0-4:0)
  \\"\\""
`;

exports[`tokenizer tokenizes 1`] = `
"number (0-8) (1:0-1:8)
  \\"01234567\\"
ignorable (8-9) (1:8-2:0)
  \\"\\\\n\\"
ident (9-16) (2:0-2:7)
  \\"foo_bar\\"
ignorable (16-17) (2:7-3:0)
  \\"\\\\n\\"
ident (17-25) (3:0-3:8)
  \\"_foo_bar\\"
ignorable (25-26) (3:8-4:0)
  \\"\\\\n\\"
ident (26-38) (4:0-4:12)
  \\"_foo_bar_123\\"
ignorable (38-39) (4:12-5:0)
  \\"\\\\n\\"
ident (39-42) (5:0-5:3)
  \\"foo\\"
ignorable (42-43) (5:3-5:4)
  \\" \\"
ident (43-46) (5:4-5:7)
  \\"bar\\"
ignorable (46-47) (5:7-6:0)
  \\"\\\\n\\"
( (47-48) (6:0-6:1)
ident (48-53) (6:1-6:6)
  \\"shrub\\"
) (53-54) (6:6-6:7)
ignorable (54-55) (6:7-7:0)
  \\"\\\\n\\"
( (55-56) (7:0-7:1)
ident (56-60) (7:1-7:5)
  \\"spam\\"
ignorable (60-61) (7:5-7:6)
  \\" \\"
ident (61-63) (7:6-7:8)
  \\"ni\\"
, (63-64) (7:8-7:9)
ignorable (64-65) (7:9-7:10)
  \\" \\"
ident (65-67) (7:10-7:12)
  \\"hi\\"
) (67-68) (7:12-7:13)
ignorable (68-69) (7:13-8:0)
  \\"\\\\n\\"
ident (69-70) (8:0-8:1)
  \\"a\\"
ignorable (70-71) (8:1-8:2)
  \\" \\"
+ (71-72) (8:2-8:3)
ignorable (72-73) (8:3-8:4)
  \\" \\"
ident (73-74) (8:4-8:5)
  \\"b\\"
ignorable (74-75) (8:5-8:6)
  \\" \\"
- (75-76) (8:6-8:7)
ignorable (76-77) (8:7-8:8)
  \\" \\"
ident (77-78) (8:8-8:9)
  \\"c\\"
ignorable (78-79) (8:9-8:10)
  \\" \\"
* (79-80) (8:10-8:11)
ignorable (80-81) (8:11-8:12)
  \\" \\"
ident (81-82) (8:12-8:13)
  \\"d\\"
ignorable (82-83) (8:13-8:14)
  \\" \\"
/ (83-84) (8:14-8:15)
ignorable (84-85) (8:15-8:16)
  \\" \\"
ident (85-86) (8:16-8:17)
  \\"f\\"
ignorable (86-87) (8:17-8:18)
  \\" \\"
^ (87-88) (8:18-8:19)
ignorable (88-89) (8:19-8:20)
  \\" \\"
ident (89-90) (8:20-8:21)
  \\"g\\"
ignorable (90-91) (8:21-9:0)
  \\"\\\\n\\"
str-start (91-92) (9:0-9:1)
  \\"\\\\\\"\\"
str-text (92-97) (9:1-9:6)
  \\"shrub\\"
str-end (97-98) (9:6-9:7)
  \\"\\\\\\"\\"
ignorable (98-99) (9:7-10:0)
  \\"\\\\n\\"
str-start (99-100) (10:0-10:1)
  \\"\\\\\\"\\"
str-text (100-103) (10:1-10:4)
  \\"ni \\"
str-expr-start (103-105) (10:4-10:6)
  \\"\${\\"
ident (105-108) (10:6-10:9)
  \\"ham\\"
str-expr-end (108-109) (10:9-10:10)
  \\"}\\"
str-text (109-114) (10:10-10:15)
  \\" spam\\"
str-end (114-115) (10:15-10:16)
  \\"\\\\\\"\\"
ignorable (115-116) (10:16-11:0)
  \\"\\\\n\\"
ident (116-119) (11:0-11:3)
  \\"foo\\"
. (119-120) (11:3-11:4)
ident (120-123) (11:4-11:7)
  \\"bar\\"
ignorable (123-124) (11:7-11:8)
  \\" \\"
... (124-127) (11:8-11:11)
ident (127-129) (11:11-11:13)
  \\"ni\\"
ignorable (129-130) (11:13-12:0)
  \\"\\\\n\\"
[ (130-131) (12:0-12:1)
true (131-135) (12:1-12:5)
, (135-136) (12:5-12:6)
ignorable (136-137) (12:6-12:7)
  \\" \\"
false (137-142) (12:7-12:12)
] (142-143) (12:12-12:13)
ignorable (143-144) (12:13-13:0)
  \\"\\\\n\\"
true (144-148) (13:0-13:4)
, (148-149) (13:4-13:5)
ignorable (149-150) (13:5-13:6)
  \\" \\"
false (150-155) (13:6-13:11)
ignorable (155-156) (13:11-14:0)
  \\"\\\\n\\"
ident (156-157) (14:0-14:1)
  \\"a\\"
ignorable (157-158) (14:1-14:2)
  \\" \\"
not (158-161) (14:2-14:5)
ignorable (161-162) (14:5-14:6)
  \\" \\"
ident (162-163) (14:6-14:7)
  \\"b\\"
ignorable (163-164) (14:7-15:0)
  \\"\\\\n\\"
ident (164-167) (15:0-15:3)
  \\"foo\\"
ignorable (167-168) (15:3-15:4)
  \\" \\"
in (168-170) (15:4-15:6)
ignorable (170-171) (15:6-15:7)
  \\" \\"
ident (171-174) (15:7-15:10)
  \\"bar\\"
ignorable (174-175) (15:10-16:0)
  \\"\\\\n\\"
end (175-175) (16:0-16:0)
  \\"\\""
`;

exports[`tokenizer tokenizes str exprs 1`] = `
"{ (0-1) (1:0-1:1)
ident (1-4) (1:1-1:4)
  \\"foo\\"
: (4-5) (1:4-1:5)
ignorable (5-6) (1:5-1:6)
  \\" \\"
str-start (6-7) (1:6-1:7)
  \\"\\\\\\"\\"
str-expr-start (7-9) (1:7-1:9)
  \\"\${\\"
ident (9-12) (1:9-1:12)
  \\"ham\\"
str-expr-end (12-13) (1:12-1:13)
  \\"}\\"
str-end (13-14) (1:13-1:14)
  \\"\\\\\\"\\"
} (14-15) (1:14-1:15)
ignorable (15-16) (1:15-2:0)
  \\"\\\\n\\"
end (16-16) (2:0-2:0)
  \\"\\""
`;
